src/tactic/rewrite_search/search.lean

```diff
@@ -0,0 +1,212 @@
+of our initial equation.
+`conf` and `rules` determine what rewrites are used to generate new graph vertices.
+Here, the format of a rewrite rule is an expression for rewriting, plus a flag for the
+direction to apply it in.
+`vertices` maps vertex.id to vertex.
+`vmap` maps vertex.pp to vertex.id, so we can quickly find collisions.
+`solving_edge` represents a solution that will prove our target equation.
+It is an edge that would connect the two trees, so `solving_edge.fr` and `solving_edge.to`
+are vertices in different trees.
+-/
+meta structure graph :=
+(conf         : config)
+(rules        : list (expr × bool))
+(vertices     : buffer vertex)
+(vmap         : native.rb_map string ℕ)
+(solving_edge : option edge)
+
+/--
+Construct a graph to search for a proof of a given equation.
+
+This graph initially contains only two disconnected vertices corresponding to the two
+sides of the equation. When `find_proof` is called, we will run a search and add
+new vertices and edges.
+-/
+meta def mk_graph (conf : config) (rules : list (expr × bool)) (eq : expr) : tactic graph :=
+do (lhs, rhs) ← split_equation eq,
+  lhs_pp ← to_string <$> tactic.pp lhs,
+  rhs_pp ← to_string <$> tactic.pp rhs,
+  let lhs_vertex : vertex := ⟨0, lhs, lhs_pp, side.L, none⟩,
+  let rhs_vertex : vertex := ⟨1, rhs, rhs_pp, side.R, none⟩,
+  return ⟨conf, rules, [lhs_vertex, rhs_vertex].to_buffer,
+          native.rb_map.of_list [(lhs_pp, 0), (rhs_pp, 1)], none⟩
+
+variables (g : graph)
+
+namespace graph
+
+/--
+A convenience wrapper around `g.vertices` that fails if the index is out of bounds.
+-/
+meta def get_vertex (i : ℕ) : tactic vertex :=
+if h : i < g.vertices.size then return $ g.vertices.read (fin.mk i h)
+else fail "invalid vertex access"
+
+/--
+Find a list of edges that connect the given edge to the root of its tree.
+The edges are returned in leaf-to-root order, while they are in root-to-leaf direction,
+so if you want them in the logical order you must reverse the returned list.
+-/
+private meta def walk_up_parents : option edge → tactic (list edge)
+| none     := return []
+| (some e) := do
+  v ← g.get_vertex e.fr,
+  edges ← walk_up_parents v.parent,
+  return (e :: edges)
+
+/--
+Returns two lists that represent a solution. The first list is a path from LHS to some
+interior vertex, the second is a path from the RHS to that interior vertex.
+-/
+private meta def solution_paths : tactic (list edge × list edge) :=
+do e ← g.solving_edge,
+  v ← g.get_vertex e.to,
+  path1 ← walk_up_parents g e,
+  path2 ← walk_up_parents g v.parent,
+  match v.side with
+  | side.L := return (path2.reverse, path1.reverse)
+  | side.R := return (path1.reverse, path2.reverse)
+  end
+
+/--
+Add the new vertex and edge to the graph, that can be proved in one step starting
+at a given vertex, with a given rewrite expression.
+For efficiency, it's important that this is the only way the graph is mutated,
```

Comment:
Ah, maybe this is the reason not to use a monad. I remember some hell dealing with reference counting objects in the monad state.

Reply:
This is the reason to use a buffer, so you can be very careful about only appending to it and never mutating it, rather than passing around lists or arrays, where it's more convenient but slower to modify things in different locations. I'm not sure about the monad part.